### DISCUSSION TITLE
Loading big GLTF efficiently

### DISCUSSION THREAD

**Discussion by avilkinnick** [vsg-dev/VulkanSceneGraph#1267] [Q&A]
Hello, I have .gltf file that represents long railway route.
View from top:
![image](https://github.com/user-attachments/assets/a559e993-cfa6-4268-8a7e-992fc212524d)
If I just load this through read_cast and add in my scene, I have around 10 FPS, what are another ways to load this?

**Comment by robertosfield** [comment]
We don't have enough information to know what to suggest as there are many different ways to improve performance, what techniques to deploy depend entirely upon the bottlenecks you are seeing.  To kick off here's some questions that will help home in on what's going on:

1. How was the glTF file generated?
2. How big is the file?  
3. Does the database contain any LOD nodes?
4. How much main and GPU memory is used up when you load the data?
5. What hardware/OS/drivers are you using?
6. What happens to performance when you zoom in?
7. Have you loaded the database in any other visualizers?  If so what performance are you seeing?

**Reply by avilkinnick** [reply]
1. glTF was created by hand from another format .dmd that is not supported by 3d editors, loaders and hard to handle, after that it was imported and exported in blender. Every node is separate.
2. .gltf file - 3.2 MB; .bin file - 5.2 MB; textures are in .png and weight 101.3 MB
3. I only yesterday started to read glTF specs and create it, so now I don't know how to add LOD to nodes
4 and 7. I loaded it in blender and glTF viewer, there are stats:

![image](https://github.com/user-attachments/assets/7804e95d-6928-452f-93d2-006031b3da68)
![image](https://github.com/user-attachments/assets/fabddba4-b9c2-47a7-88ca-6fedfb374bd6)

5. Linux, Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, NVIDIA GeForce GTX 970
6. When I zoom, performance doesn't change

**Comment by robertosfield** [comment]
I'm not familiar with the DMD format, but you may find it better to just write a DMD loader for the VSG rather than map to glTF.

The model size sounds pretty small, not enough to result in a 10fps framerate even on an old card like a GTX 970, so I'd suspect something about it is tripping up performance.

When you say every node was separate, how many nodes are there in the scene?

The VSG provides a wide range of features to help performance on very large models - culling nodes like vsg::CullGroup/CullGroup, level of details nodes like vsg::LOD and to support databases paging vsg::PagedLOD.  The later is what enables applications to scale to rendering the whole world at a solid 60fps.

How familiar are you with scene graphs/concepts in graphics rendering?

**Reply by avilkinnick** [reply]
Hello, after some break, I am trying to load this scene again. I was using PagedLOD earlier, but I don't quite understand how to use them with .gltf where only textures are stored separate but all vertex data is inside one .bin file, besides this, I also don't know, should I use some library like tinyGltf to split gltf or I can somehow use node returned by vsg::read_cast. And, about scene graphs, I understand the idea, but now all route's nodes are separate, I don't know how I can logically organize them in a tree
